src/sentry/static/sentry/app/views/eventsV2/table/tableView.tsx

```diff
@@ -212,48 +216,131 @@ class TableView extends React.Component<TableViewProps> {
+      return this.renderBrowserExportButton(canEdit);
+    } else {
+      return (
+        <Feature
+          features={['organizations:data-export']}
+          renderDisabled={() => this.renderBrowserExportButton(canEdit)}
+        >
+          {this.renderAsyncExportButton(canEdit)}
+        </Feature>
+      );
+    }
+  }
+
+  handleDownloadAsCsv = () => {
+    const {organization, title, eventView, tableData} = this.props;
+    trackAnalyticsEvent({
+      eventKey: 'discover_v2.results.download_csv',
+      eventName: 'Discoverv2: Download CSV',
+      organization_id: parseInt(organization.id, 10),
+    });
+    downloadAsCsv(tableData, eventView.getColumns(), title);
+  };
+
+  renderBrowserExportButton(canEdit: boolean) {
+    const disabled = this.props.isLoading || canEdit === false;
+    const onClick = disabled ? undefined : this.handleDownloadAsCsv;
+
+    return (
+      <HeaderButton
+        disabled={disabled}
+        onClick={onClick}
+        data-test-id="grid-download-csv"
+      >
+        <IconDownload size="xs" />
+        {t('Export Page')}
+      </HeaderButton>
+    );
+  }
+
+  renderAsyncExportButton(canEdit: boolean) {
+    const {isLoading, location} = this.props;
+    const disabled = isLoading || canEdit === false;
+    return (
+      <HeaderDownloadButton
+        payload={{
+          queryType: ExportQueryType.Discover,
+          queryInfo: location.query,
+        }}
+        disabled={disabled}
+      >
+        <IconDownload size="xs" />
+        {t('Export All')}
+      </HeaderDownloadButton>
+    );
+  }
+
+  renderEditButton(canEdit: boolean) {
+    const onClick = canEdit ? this.handleEditColumns : undefined;
+    return (
+      <HeaderButton disabled={!canEdit} onClick={onClick} data-test-id="grid-edit-enable">
+        <IconEdit size="xs" />
+        {t('Edit Columns')}
+      </HeaderButton>
+    );
+  }
+
   render() {
-    const {
-      isLoading,
-      error,
-      location,
-      tableData,
-      eventView,
-      title,
-      organization,
-    } = this.props;
+    const {isLoading, error, location, tableData, eventView} = this.props;
 
     const columnOrder = eventView.getColumns();
     const columnSortBy = eventView.getSorts();
 
     return (
       <GridEditable
-        editFeatures={['organizations:discover-query']}
-        noEditMessage={t('Requires discover query feature.')}
         isLoading={isLoading}
         error={error}
         data={tableData ? tableData.data : []}
         columnOrder={columnOrder}
         columnSortBy={columnSortBy}
+        title={t('Results')}
         grid={{
           renderHeadCell: this._renderGridHeaderCell as any,
           renderBodyCell: this._renderGridBodyCell as any,
           onResizeColumn: this._resizeColumn as any,
           renderPrependColumns: this._renderPrependColumns as any,
           prependColumnWidths: ['40px'],
         }}
+        headerButtons={this.renderHeaderButtons}
```

Comment:
There's a lot of LOC to render the header, does it make sense for this to be in its own component?

Reply:
Yes. I wanted to lift and shift first.